### ISSUE TITLE
step number over-counted compare to epochs number, leads to early termination

### ISSUE THREAD

**Issue by zhuliyi0** [bghira/SimpleTuner#160]
I think the gra accu steps is not considered when calcualting num epochs from max train steps. The test run I ran was 2 accu steps and should be 1 step per epoch, but the global step count end up twice as many as epoch count, and the run ended at half the max steps.

Other earlier runs had similar issue.

**Comment by zhuliyi0** [comment]
If num_train_epoch is used instead of max_train_steps, the right amount of epoch is trained, but total steps showed up as twice as needed.

**Comment by bghira** [comment]
```py
    # We need to recalculate our total training steps as the size of the training dataloader may have changed.
    num_update_steps_per_epoch = math.ceil(
        len(train_dataloader) / args.gradient_accumulation_steps
    )
    if overrode_max_train_steps:
        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
    # Afterwards we recalculate our number of training epochs
    args.num_train_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)
```

there are a few potential issues to look at here;

the first one is that it seems like most pytorch dataloaders are designed to return the length of all possible batches that the dataloader could create. in other words, it should be `floor(dataloader_length / batch_size)` to ensure we capture an integer value of the number of batches, assuming we'll discard any images that don't fit into the batch size.

the `overrode_max_train_steps` is engaged if `max_train_steps=0`.

cant state much more without some log output to help piece it together, but i've added a commit to at least address that part.

**Comment by bghira** [comment]
basically, the bug is that the batch size actually wasn't being taken into consideration when calculating the training steps.

the length of the dataset was all samples in the dataset, and dividing that by the number of gradient accumulation steps is incorrect, as it means we'll not account for the fact that we group samples.
